mlir/lib/Analysis/Presburger/Barvinok.cpp

```diff
@@ -147,6 +149,314 @@ GeneratingFunction mlir::presburger::detail::unimodularConeGeneratingFunction(
+    a2 = FracMatrix(
+        remainder.getSubMatrix(0, numIneqs - numVars - 1, 0, numVars - 1));
+    b2c2 = FracMatrix(remainder.getSubMatrix(0, numIneqs - numVars - 1, numVars,
+                                             numVars + numSymbols));
+
+    // Find the vertex, if any, corresponding to the current subset of
+    // inequalities.
+    std::optional<ParamPoint> vertex =
+        solveParametricEquations(FracMatrix(subset)); // d x (p+1)
+
+    if (!vertex)
+      continue;
+    // If this subset corresponds to a vertex, store it.
+    vertices.push_back(*vertex);
+    // subsets.push_back(subset);
+
+    // Let the current vertex be [X | y], where
+    // X represents the coefficients of the parameters and
+    // y represents the constant term.
+
+    // The region (in parameter space) where this vertex is active is given
+    // by substituting the vertex into the *remaining* inequalities of the
+    // polytope (those which were not collected into `subset`), i.e.,
+    // [A2 | B2 | c2].
+    // Thus, the coefficients of the parameters after substitution become
+    // (A2 • X + B2)
+    // and the constant terms become
+    // (A2 • y + c2).
+    // The region is therefore given by
+    // (A2 • X + B2) p + (A2 • y + c2) ≥ 0
+    // This is equivalent to A2 • [X | y] + [B2 | c2]
+    // Thus we premultiply [X | y] with each row of A2
+    // and add each row of [B2 | c2].
+    FracMatrix activeRegion(numIneqs - numVars, numSymbols + 1);
+    for (unsigned i = 0; i < numIneqs - numVars; i++) {
+      activeRegion.setRow(i, vertex->preMultiplyWithRow(a2.getRow(i)));
+      activeRegion.addToRow(i, b2c2.getRow(i), 1);
+    }
+
+    // If any row of activeRegion is all-zero, that means that the
+    // corresponding inequality in `remaining` is *also* satisfied by the
+    // vertex for any values of the parameters. Thus we remove it from
+    // activeRegion and add that inequality to `subset`.
+    // Note that if the row is not all-zero, it is still possible for the
+    // inequality to be satisfied by the corresponding vertex *in some subset of
+    // the parameter space*. However, since this subset is defined by an
+    // equality, it is not full-dimensional and we can therefore ignore it.
+    for (int i = activeRegion.getNumRows() - 1; i >= 0; --i) {
+      if (llvm::any_of(activeRegion.getRow(i),
+                       [&](Fraction f) { return f != 0; }))
+        continue;
+      subset.appendExtraRow(remainder.getRow(i));
+      activeRegion.removeRow(i);
+    }
+
+    // We convert the representation of the active region to an integers-only
+    // form so as to store it as a PresburgerSet.
+    IntMatrix activeRegionNorm = activeRegion.normalizeRows();
+    IntegerPolyhedron activeRegionRel = IntegerPolyhedron(
+        PresburgerSpace::getRelationSpace(0, numSymbols, 0, 0));
+    for (unsigned i = 0, e = activeRegion.getNumRows(); i < e; ++i)
+      activeRegionRel.addInequality(activeRegionNorm.getRow(i));
```

Comment:
This seems to be a very common operation, so you should add a constructor the IntegerPolyhedron for this